pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -178,55 +239,54 @@ func (p *ecrProvider) Enabled() bool {
+	if getter, ok := f.cache[region]; ok {
+		return getter, nil
+	}
+	getter, err := newECRTokenGetter(region)
+	if err != nil {
+		klog.Errorf("while getting creating token getter for region %v %v", region, err)
+		return nil, err
+	}
+	f.cache[region] = getter
+	return f.cache[region], nil
 }
 
-// LazyProvide implements DockerConfigProvider.LazyProvide. Should never be called.
-func (p *ecrProvider) LazyProvide() *credentialprovider.DockerConfigEntry {
-	return nil
+// The canonical implementation
+type ecrTokenGetter struct {
+	svc *ecr.ECR
 }
 
-// Provide implements DockerConfigProvider.Provide, refreshing ECR tokens on demand
-func (p *ecrProvider) Provide() credentialprovider.DockerConfig {
-	cfg := credentialprovider.DockerConfig{}
+// GetAuthorizationToken gets the ECR authorization token using the ECR API
+func (p *ecrTokenGetter) GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput) (*ecr.GetAuthorizationTokenOutput, error) {
+	return p.svc.GetAuthorizationToken(input)
+}
 
-	// TODO: fill in RegistryIds?
-	params := &ecr.GetAuthorizationTokenInput{}
-	output, err := p.getter.GetAuthorizationToken(params)
-	if err != nil {
-		klog.Errorf("while requesting ECR authorization token %v", err)
-		return cfg
-	}
-	if output == nil {
-		klog.Errorf("Got back no ECR token")
-		return cfg
-	}
+type cacheEntry struct {
+	expiresAt   time.Time
+	credentials credentialprovider.DockerConfigEntry
+	registry    string
+}
 
-	for _, data := range output.AuthorizationData {
-		if data.ProxyEndpoint != nil &&
-			data.AuthorizationToken != nil {
-			decodedToken, err := base64.StdEncoding.DecodeString(aws.StringValue(data.AuthorizationToken))
-			if err != nil {
-				klog.Errorf("while decoding token for endpoint %v %v", data.ProxyEndpoint, err)
-				return cfg
-			}
-			parts := strings.SplitN(string(decodedToken), ":", 2)
-			user := parts[0]
-			password := parts[1]
-			entry := credentialprovider.DockerConfigEntry{
-				Username: user,
-				Password: password,
-				// ECR doesn't care and Docker is about to obsolete it
-				Email: "not@val.id",
-			}
-
-			klog.V(3).Infof("Adding credentials for user %s in %s", user, p.region)
-			// Add our config entry for this region's registry URLs
-			cfg[p.regionURL] = entry
-
-		}
-	}
-	return cfg
+// ecrExpirationPolicy implements ExpirationPolicy from client-go.
+type ecrExpirationPolicy struct{}
+
+// stringKeyFunc returns the cache key as a string
+func stringKeyFunc(obj interface{}) (string, error) {
+	key := obj.(*cacheEntry).registry
+	return key, nil
+}
+
+// IsExpired checks if the ECR credentials are expired. Credentials expire at
+// 1/2 of their original requested window.
+func (p *ecrExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	ecrExpiresAt := entry.Obj.(*cacheEntry).expiresAt
+	validWindow := ecrExpiresAt.Sub(entry.Timestamp)
+	expirationTime := ecrExpiresAt.Add(-1 * validWindow / time.Duration(2))
```

Comment:
I think this could read better like this:
```go
expirationTime := ecrExpiresAt.Sub(validWindow / 2)
return expirationTime.After(time.Now())
```

Reply:
Sub requires a duration.

Thoughts on:
```go
halfLife := ecrExpiresAt.Sub(entry.Timestamp) / time.Duration(2)
return ecrExpiresAt.Add(halfLife * -1).Before(time.Now())
```

Reply:
Oh, I see! Yeah that seems better :)